### ISSUE TITLE
full_like, ones_like, zeros_like should retain subclasses

### ISSUE THREAD

**Issue by gerritholl** [pydata/xarray#2542]
#### Code Sample,

```python
# Your code here
import numpy
import xarray

class MyDataArray(xarray.DataArray):
    pass

da = MyDataArray(numpy.arange(5))
da2 = xarray.zeros_like(da)
print(type(da), type(da2))

```
#### Problem description

I would expect that `type(da2) is type(da)`, but this is not the case.  The type of `da` is always `<class 'xarray.core.dataarray.DataArray'>`.  Rather, the output of this script is:

```<class '__main__.MyDataArray'> <class 'xarray.core.dataarray.DataArray'>```

#### Expected Output

I would hope as an output:

```<class '__main__.MyDataArray'> <class '__main__.MyDataArray'>```

In principle changing this could break people's code, so if a change is implemented it should probably be through an optional keyword argument to the `full_like`/`ones_like`/`zeros_like` family.

#### Output of ``xr.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.7.0.final.0
python-bits: 64
OS: Linux
OS-release: 2.6.32-754.el6.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: en_GB.UTF-8

xarray: 0.10.7
pandas: 0.23.2
numpy: 1.15.2
scipy: 1.1.0
netCDF4: 1.4.0
h5netcdf: 0.6.1
h5py: 2.8.0
Nio: None
zarr: None
bottleneck: 1.2.1
cyordereddict: None
dask: 0.18.1
distributed: 1.22.0
matplotlib: 3.0.0
cartopy: 0.16.0
seaborn: 0.9.0
setuptools: 39.2.0
pip: 18.0
conda: None
pytest: 3.2.2
IPython: 6.4.0
sphinx: None

</details>

**Comment by shoyer** [comment]
Are there other functions where we don't do this?

I would be happy to accept a fix here. I don't think we need a few keyword argument -- I expect most subclass users would consider this a bug rather than a feature.

We might consider adding an official subclassing APIs like pandas:
https://pandas.pydata.org/pandas-docs/stable/extending.html#extending-subclassing-pandas

**Comment by TomNicholas** [comment]
See #3980 - this issue should be solved as part of #3980, but see that issue for more general discussion of subclassing.

**Comment by max-sixty** [comment]
Shall we fold this into #3980? (i.e. I don't think we need an issue for enabling subclassing for each method :) )
